### PR TITLE
fixed honoring custom repr for NamedTuple if assigned by partialmethod

### LINKED ISSUE #14465: `display` does not honor custom `__repr__` for `NamedTuple` when defined by `partialmethod`
The issue occurs when the custom `__repr__` is defined via `partialmethod`:

```python
from functools import partialmethod
from typing import NamedTuple

def my_pprint(obj, **options):
    return "Hello World"

def add_custom_pprint(cls=None, **options):
    if cls is None:
        def decorator(cls):
            return add_custom_pprint(cls, **options)
        return decorator
    cls.__repr__ = partialmethod(my_pprint, **options)
    return cls

@add_custom_pprint
class MyNamedTuple(NamedTuple):
    x: int
    y: int

x = MyNamedTuple(1, 2)
print(x)  # "Hello World"
display(x)  # (1, 2)
```

<details><summary> without partialmethod, both display and print give the same result: </summary>

```python
from typing import NamedTuple

def my_pprint(obj, **options):
    return "Hello World"

def add_custom_pprint(cls=None, **options):
    if cls is None:
        def decorator(cls):
            return add_custom_pprint(cls, **options)
        return decorator
    cls.__repr__ = my_pprint
    return cls

@add_custom_pprint
class MyNamedTuple(NamedTuple):
    x: int
    y: int

x = MyNamedTuple(1, 2)
print(x)  # "Hello World"
display(x)   # "Hello World"
```

</details>

### PR DESCRIPTION
**Pull request by randolf-scholz**
Fixes #14465.

- added test `test_custom_repr_namedtuple_partialmethod`

### CODE REVIEW

File: IPython/core/tests/test_formatters.py [hunk 1]
```diff
@@ -103,7 +103,7 @@ def set_fp(p):
 
 def test_for_type():
     f = PlainTextFormatter()
-
```
Comment (randolf-scholz) [inline IPython/core/tests/test_formatters.py:106]: I think these got automatically added by the pre-commit.

Comment (Carreau) [reply IPython/core/tests/test_formatters.py:106]: no worries.

File: IPython/lib/pretty.py [hunk 2]
```diff
@@ -406,8 +406,16 @@ def pretty(self, obj):
...
-                                and callable(cls.__dict__.get('__repr__')):
+                        if (
+                            cls is not object
+                            # check if cls defines __repr__
+                            and "__repr__" in cls.__dict__
+                            # check if __repr__ is callable.
+                            # Note: we need to test getattr(cls, '__repr__')
+                            #   instead of cls.__dict__['__repr__']
+                            #   in order to work with descriptors like partialmethod,
+                            and callable(getattr(cls, "__repr__", None))
```
Comment (Carreau) [inline IPython/lib/pretty.py:417]: I think we have a `_safe_getattr` that tries to detect mock objects that always return mock objects that always have existing methods.

**Comment by randolf-scholz** [comment]
Seems like the two failing CI pipelines are not related to this PR, they fail on others too: https://github.com/ipython/ipython/pull/14451, https://github.com/ipython/ipython/pull/14460

**Comment by Carreau** [comment]
yeah, don't worry about the failures.

**Pull request merged**